OPAL/si/src/main/scala/org/opalj/fpcf/scheduling/MaximumPhaseScheduling.scala

```diff
@@ -98,7 +102,60 @@ abstract class MultiplePhaseScheduling extends SchedulingStrategy {
             computePhase(ps, phaseAnalyses, remainingAnalyses)
         }
 
-        schedule
+        if (!config.getBoolean(DisableCleanup)) calculateDeletions(schedule, ps, config) else schedule
+    }
+
+    private def calculateDeletions[A](
+        schedule: List[PhaseConfiguration[A]],
+        ps:       PropertyStore,
+        config:   Config
+    ): List[PhaseConfiguration[A]] = {
+        def keyByName(name: String): SomePropertyKey = {
+            val max = PropertyKey.maxId
+            var i = 0
+            while (i <= max) {
+                if (PropertyKey.name(i) == name) return PropertyKey.key(i)
+                i += 1
+            }
+            throw new IllegalArgumentException(s"Unknown property name: $name")
+        }
+        def getSetForProperty(propertyKey: String): Set[Int] = {
+            Option(
+                config.getString(propertyKey)
+            ).toSeq.flatMap(_.split(",")).filter(_.nonEmpty).map(name => keyByName(name).id).toSet
+        }
+        val pkToKeep = getSetForProperty(PropertiesToKeepKey)
+        val pkToDelete = getSetForProperty(PropertiesToRemoveKey)
+        val properties: Set[Int] =
+            schedule.iterator.flatMap(
+                _.propertyKinds.propertyKindsComputedInThisPhase.map(_.id)
+            ).toSet
+
+        val modifiedSchedule: ListBuffer[PhaseConfiguration[A]] = ListBuffer.from(schedule)
+
+        var usedDependencies: Set[Int] = Set.empty
+        var alreadyDeleted: Set[Int] = Set.empty
+        val lastIndex = modifiedSchedule.length - 1
+        var index = lastIndex
+        while (index >= 0) {
+            val elem = modifiedSchedule(index)
+            val currentSchedule = elem.scheduled
+            usedDependencies ++= currentSchedule.iterator.flatMap(_.uses(ps).iterator.map(_.pk.id)).toSet
+            usedDependencies ++= pkToKeep
+            val remaining: Set[Int] = {
+                if (index == lastIndex) {
+                    (properties -- pkToKeep) -- alreadyDeleted
+                } else
+                    (properties -- usedDependencies) -- alreadyDeleted
+            }
+
+            alreadyDeleted ++= remaining
+            index -= 1
+        }
+        modifiedSchedule(lastIndex) = modifiedSchedule(lastIndex).copy(
+            toDelete = (modifiedSchedule(lastIndex).toDelete ++ pkToDelete) -- pkToKeep
```

Comment:
This looks like it would only compute PKs to be deleted after the last phase, which is clearly not what we want